File: java/leetcode/src/main/java/com/leetcode/tree/Most_Frequent_Subtree_Sum_508.java
package com.leetcode.tree;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Most_Frequent_Subtree_Sum_508 {
    public int[] findFrequentTreeSum(TreeNode root) {
        int max = 0;
        Map<Integer, Integer> map = new HashMap<>();
        sum(root, map);
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() > max) {
                max = entry.getValue();
            }
        }

        List<Integer> list = new ArrayList<>();
        for(Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if(entry.getValue() == max) {
                list.add(entry.getKey());
            }
        }

        int[] res = new int[list.size()];
        int cnt = 0;
        for (Integer n : list) {
            res[cnt++] = n;
        }

        return res;
    }

    int sum(TreeNode node,  Map<Integer, Integer> map) {
        if (node == null) return 0;
        int sum = sum(node.left, map) + sum(node.right, map) + node.val;
        if (map.containsKey(sum)) {
            map.put(sum, map.get(sum) + 1);
        } else {
            map.put(sum, 1);
        }
        return sum;
    }
}

File: js/algrithm/tree/Most_Frequent_Subtree_Sum_508.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function(root) {
    const map = new Map([])
    const dfs = function(node) {
        if (!node) return 0
        const left = dfs(node.left)
        const right = dfs(node.right)
        const sum = node.val + left + right
        map.set(sum, (map.get(sum) || 0) + 1)
        return sum
    }
    dfs(root)
    const arr = [...map].sort((a, b) => b[1] - a[1])
    const res = []
    for (let i = 0; i < arr.length; i++) {
        if (arr[i][1] === arr[0][1]) {
            res.push(arr[i][0])
        } else {
            break
        }
    }
    return res
    
};